### PR TITLE
Update to Chef 12.5 custom resources

### LINKED ISSUE #217: Require Chef Infra Client 12.5+
We set the requirement of this cookbook to Chef Infra Client 11+ several years ago. That makes this cookbook rather difficult to maintain as it cannot be written as a simpler custom resource and we have to be incredibly defensive in our resource usage. Let's bump to 12.5+ now so we can use customer resources.

### PR DESCRIPTION
**Pull request by detjensrobert**
Signed-off-by: Robert Detjens <detjensrobert@osuosl.org>

### Description

Instead of using Chef 11 resource/providers, switch to Chef 12 custom resources. This also enables `unified_mode` for Chef 17 compatibility.

### Issues Resolved

Fixes #217.

### Check List

- [x] All tests pass. See <https://github.com/chef-cookbooks/community_cookbook_documentation/blob/master/TESTING.MD>
- [ ] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable
- [x] All commits have been signed for the Developer Certificate of Origin. See <https://github.com/chef-cookbooks/community_cookbook_documentation/blob/master/CONTRIBUTING.MD>

### CODE REVIEW

File: .rubocop.yml [hunk 1]
```diff
@@ -1,2 +1,2 @@
 AllCops:
-  TargetChefVersion: 11
\ No newline at end of file
+  TargetChefVersion: 12.latest
```
Comment (tas50) [inline .rubocop.yml:2]:
```suggestion
  TargetChefVersion: 12.5
```

File: attributes/default.rb [hunk 2]
```diff
@@ -2,7 +2,7 @@
 # Cookbook:: chef_client_updater
 # Attributes:: default
 #
-# Copyright:: 2016-2020, Chef Software Inc.
+# Copyright:: 2016-2021, Chef Software Inc.
```
Comment (tas50) [inline attributes/default.rb:5]:
```suggestion
# Copyright:: Chef Software Inc.
```

File: libraries/matchers.rb [hunk 3]
```diff
@@ -1,4 +1,4 @@
-# cookstyle: disable ChefModernize/DefinesChefSpecMatchers
+# cookstyle: disable Chef/Modernize/DefinesChefSpecMatchers
 if defined?(ChefSpec)
```
Comment (tas50) [inline libraries/matchers.rb:2]: These should get deleted at this point

File: metadata.rb [hunk 4]
```diff
@@ -5,13 +5,13 @@
...
+chef_version '>= 12.5' if respond_to?(:chef_version) # cookstyle: disable Chef/Modernize/RespondToInMetadata
 
 %w(amazon centos debian mac_os_x opensuseleap oracle redhat scientific solaris2 suse ubuntu windows aix).each do |os|
   supports os
 end
 
-unless defined?(Ridley) # cookstyle: disable ChefModernize/RespondToInMetadata
-  source_url 'https://github.com/chef-cookbooks/chef_client_updater' if respond_to?(:source_url) # cookstyle: disable ChefModernize/RespondToInMetadata
-  issues_url 'https://github.com/chef-cookbooks/chef_client_updater/issues' if respond_to?(:issues_url) # cookstyle: disable ChefModernize/RespondToInMetadata
+unless defined?(Ridley) # cookstyle: disable Chef/Modernize/RespondToInMetadata
```
Comment (tas50) [inline metadata.rb:14]: I think the unless Ridley was a chef 11 thing that can go away but @lamont-granquist would know best here

File: resources/default.rb [hunk 5]
```diff
@@ -2,8 +2,8 @@
 # Cookbook:: chef_client_updater
 # Resource:: updater
 #
-# Copyright:: 2016-2018, Will Jordan
-# Copyright:: 2016-2020, Chef Software Inc.
+# Copyright:: 2016-2021, Will Jordan
+# Copyright:: 2016-2021, Chef Software Inc.
```
Comment (tas50) [inline resources/default.rb:6]:
```suggestion
# Copyright:: Chef Software Inc.
```

File: resources/default.rb [hunk 6]
```diff
@@ -18,32 +18,691 @@
...
+        clear_sources true if respond_to?(:clear_sources)
+        options "--source #{new_resource.rubygems_url}" if respond_to?(:options)
+      end
+    end
+
+    require 'mixlib/versioning'
+  end
+
+  def update_rubygems
+    compatible_rubygems_versions = '>= 2.6.11'
```
Comment (tas50) [inline resources/default.rb:84]: it's worth checking if 12.5 comes with 2.6.11+ already

Comment (tas50) [reply resources/default.rb:84]: Nopt it has 2.4.8

File: resources/default.rb [hunk 7]
```diff
@@ -18,32 +18,691 @@
...
+    target_version = '2.7.8' # should be bumped to latest 2.x, but ruby < 2.3.0 support is necessary
+    nodoc_rubygems_versions = '>= 3.0'
+
+    rubygems_version = Gem::Version.new(Gem::VERSION)
+    Chef::Log.debug("Found gem version #{rubygems_version}. Desired version is #{compatible_rubygems_versions}")
+    return if Gem::Requirement.new(compatible_rubygems_versions).satisfied_by?(rubygems_version)
+
+    # only rubygems >= 1.5.2 supports pinning, and we might be coming from older versions
+    pin_rubygems_range = '>= 1.5.2'
+    pin = Gem::Requirement.new(pin_rubygems_range).satisfied_by?(rubygems_version)
```
Comment (tas50) [inline resources/default.rb:94]: Chef 12.5 comes with 2.4.8 so pinning is always supported now. We can remove all this logic

File: resources/default.rb [hunk 8]
```diff
@@ -18,32 +18,691 @@
...
+
+    require 'mixlib/install'
+  end
+
+  def load_mixlib_versioning
+    require 'mixlib/versioning'
+  rescue LoadError
+    Chef::Log.info('mixlib-versioning gem not found. Installing now')
+    chef_gem 'mixlib-versioning' do
+      compile_time true if respond_to?(:compile_time) # cookstyle: disable Chef/Modernize/RespondToCompileTime
```
Comment (tas50) [inline resources/default.rb:73]: These can all become `compile_time true` now that we have Chef 12.1+

File: resources/default.rb [hunk 9]
```diff
@@ -18,32 +18,691 @@
...
+  # @return nil when product is not installed
+  #
+  def current_version
+    case new_resource.product_name
+    when 'chef', 'angrychef', 'cinc', 'angrycinc'
+      node['chef_packages']['chef']['version']
+    when 'chefdk'
+      versions = Mixlib::ShellOut.new('chef -v').run_command.stdout # cookstyle: disable Chef/Modernize/ShellOutHelper
+      # There is a verbiage change in newer version of Chef Infra
+      version = versions.match(/(ChefDK Version(.)*:)\s*([\d.]+)/i) || versions.match(/(Chef Development Kit Version(.)*:)\s*([\d.]+)/i)
```
Comment (tas50) [inline resources/default.rb:141]: We can remove all the chefdk junk here now

**Review by jeffbyrnes — APPROVED** [review]

**Comment by tas50** [comment]
Sadly I'm going to have to close this out. We have a large number of folks trying to use this today to migrate off pre-12.5 releases. It's sad, but the reality we have.

**Comment by jeffbyrnes** [comment]
@tas50 wouldn't those folks be able to stick with an older version of this cookbook to do those migrations?

**Comment by detjensrobert** [comment]
What's the plan for `unified_mode` Chef 17/18 compatibility with this cookbook if this PR is closed? Does `unified_mode` work with the legacy LWRPs?

**Comment by tas50** [comment]
@detjensrobert We can make it work only on systems with unified mode.

**Comment by jeffbyrnes** [comment]
Ok, makes sense! Thanks.

**Comment by lamont-granquist** [comment]
`unified_mode false if defined?(:unified_mode)` is probably the way to solve this

**Pull request closed without merging**